internal/wasmldr/src/loader/pki.rs

```diff
@@ -0,0 +1,110 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{anyhow, Result};
+use pkcs8::{AlgorithmIdentifier, ObjectIdentifier, PrivateKeyInfo, SubjectPublicKeyInfo};
+use zeroize::Zeroizing;
+
+use sec1::EcPrivateKey;
+use x509::der::{Decodable, Encodable};
+
+use const_oid::db::rfc5912::{
+    ECDSA_WITH_SHA_256, ECDSA_WITH_SHA_384, ID_EC_PUBLIC_KEY as ECPK, SECP_256_R_1 as P256,
+    SECP_384_R_1 as P384,
+};
+
+const ES256: AlgorithmIdentifier<'static> = AlgorithmIdentifier {
+    oid: ECDSA_WITH_SHA_256,
+    parameters: None,
+};
+
+const ES384: AlgorithmIdentifier<'static> = AlgorithmIdentifier {
+    oid: ECDSA_WITH_SHA_384,
+    parameters: None,
+};
+
+pub trait PrivateKeyInfoExt {
+    /// Generates a keypair
+    ///
+    /// Returns the DER encoding of the `PrivateKeyInfo` type.
+    fn generate(oid: ObjectIdentifier) -> Result<Zeroizing<Vec<u8>>>;
+
+    /// Get the public key
+    ///
+    /// This function creates a `SubjectPublicKeyInfo` which corresponds with
+    /// this private key. Note that this function does not do any cryptographic
+    /// calculations. It expects that the `PrivateKeyInfo` already contains the
+    /// public key.
+    fn public_key(&self) -> Result<SubjectPublicKeyInfo<'_>>;
+
+    /// Get the default signing algorithm for this `SubjectPublicKeyInfo`
+    fn signs_with(&self) -> Result<AlgorithmIdentifier<'_>>;
+
+    /// Signs the body with the specified algorithm
+    ///
+    /// Note that the signature is returned in its encoded form as it will
+    /// appear in an X.509 certificate or PKCS#10 certification request.
+    fn sign(&self, body: &[u8], algo: AlgorithmIdentifier<'_>) -> Result<Vec<u8>>;
+}
+
+impl<'a> PrivateKeyInfoExt for PrivateKeyInfo<'a> {
+    fn generate(oid: ObjectIdentifier) -> Result<Zeroizing<Vec<u8>>> {
+        let rand = ring::rand::SystemRandom::new();
+
+        let doc = match oid {
+            P256 => {
+                use ring::signature::{EcdsaKeyPair, ECDSA_P256_SHA256_ASN1_SIGNING as ALG};
```

Comment:
Why not just import `EcdsaKeyPair` at top-level, since it's used in both branches?
Also, not sure there is much benefit from renaming the other import to `ALG`, could also have just been imported at top-level and used here directly with default name

Reply:
1. I think Arm uses RSA, which will be a different code flow. We are positioned now for that. Also, this code is copy/paste from steward.

2. The renaming like this prevents awkward line-wrapping.